Drivers: sensors: Add support for AMS AS621x series

jtbaumann opened the pull request:
Added support for the AMS AS621x series of temperature sensors as a
variant of the TI TMP108 temperature sensor.

Signed-off-by: Jared Baumann <jared.baumann8@t-mobile.com>

catch-twenty-two commented:
@gmarull @MaureenHelm @mbolivar rather than submitting a new driver for this chip, we wanted to reuse the TMP108 code I had submitted already since 90% of the functionality of AMS AS621x is the same. (just different registers) We weren't sure what the best way to do this was, but thought this route made sense. Please let us know what you think.

MaureenHelm commented on drivers/sensor/tmp108/tmp108.c:
```diff
@@ -202,13 +199,18 @@ static int tmp108_attr_set(const struct device *dev,
 
 	switch ((int) attr) {
 	case SENSOR_ATTR_HYSTERESIS:
+#if DT_PROP(DT_INST(0, ti_tmp108), variant_as621x)
+		LOG_WRN("AS621x Series lacks Hysterisis setttings");
+		break;
+#endif
```
This won't work if there are multiple instances of the driver. Add an element to `struct tmp108_config` so you can look up the value on a per-instance basis.

catch-twenty-two replied:
Thanks @MaureenHelm will add this

catch-twenty-two replied:
@MaureenHelm I couldn't really think of an easy way to do this, so I just ended up requiring the user to configure this in the Kconfig by indicating whether they are compiling for a board with the as621 or the tmp108. At this point it would be a lot of work to allow both chipsets to operate on the same board using the same driver, so I think this is a good compromise.

MaureenHelm reviewed (changes requested):
Please rebase and remove the merge commit

gmarull commented on dts/bindings/sensor/ti,tmp108.yaml:
```diff
@@ -16,3 +17,8 @@ properties:
         Identifies the ALERT signal, which is active-low open drain when
         produced by the sensor.
+  
+    variant-as621x:
```
I think the right way to solve this case is to make the driver support two compatibles, factorizing common code. cc: @mbolivar-nordic

mbolivar-nordic replied:
Agreed

catch-twenty-two replied:
thanks, @gmarull and @mbolivar-nordic.  Sorry just getting back from a long break on this, so that's what we attempted to do.  1 register set is included if there is an AMS621 included in the device tree otherwise the register set from the TMP108 is included. Just wanted to make sure the approach is sound before continuing on with it.

catch-twenty-two replied:
@gmarull @mbolivar-nordic done, please take a look when you have a few minutes.

mbolivar-nordic replied:
Did you forget to push, @catch-twenty-two ? I see no changes.

catch-twenty-two replied:
Hi @mbolivar-nordic , I did make a bunch of changes. The drivers share a common .h file and the few registers definitions that are different are included depending on which driver is chosen in the KConfig example (line 22):
```
#if CONFIG_AS621x
#include "as621x_defs.h"
#else
#include "tmp108_defs.h"
#endif
```

mbolivar-nordic replied:
> Did you have something else in mind?

Yes. Take another look at @gmarull's advice (emphasis mine):

> I think the right way to solve this case is to **make the driver support two compatibles**, factorizing common code. 

I.e., you need two different devicetree bindings, one for each compatible. The driver should support **both** compatibles, and importantly both **at the same time**.

catch-twenty-two replied:
@mbolivar-nordic Thanks for your help, I have reworked everything so this driver now supports multiple instances of both the tmp108 and as6212 on the same board, if my approach looks ok Ill go ahead and clean everything up.

catch-twenty-two replied:
@gmarull fyi

mbolivar-nordic replied:
>  I have reworked everything so this driver now supports multiple instances of both the tmp108 and as6212 on the same board, if my approach looks ok Ill go ahead and clean everything up.

Did you forget to push something? I still see the problematic `variant-as621x` property.

Regardless of whether you create two drivers or create one driver that supports two compatibles, you still need two devicetree bindings.

Remember: a compatible property says what type of hardware the node represents. Having compatible be "ti,tmp108", but then adding a property that says "just kidding, this is actually an as621x", is not the right thing to do. Two parts --> two compatibles.

catch-twenty-two replied:
Ah okay, thanks for the information, I didn't realize that, will make the change

catch-twenty-two commented:
> Please rebase and remove the merge commit

Done

JordanYates commented:
How similar are these two sensors, beyond the fact that they both have a single config register and a result register?
* The sample rates aren't all the same
* The output data format is not the same
* AS621 has no hysteresis settings
* TMP108 has no consecutive fault settings

You mention `Enable driver for the TMP108 temperature sensor and it's variant the AMS AS621`, but the changes seem larger than a variant to me (I would expect accuracy changes across a variant, not different functionality).

mmahadevan108 commented on drivers/sensor/tmp108/tmp108.h:
```diff
@@ -13,33 +14,30 @@
-#define TI_TMP108_REG_TEMP		0x00   /** Temperature register */
-#define TI_TMP108_REG_CONF		0x01   /** Configuration register */
+#define TI_TMP108_REG_TEMP			0x00   /** Temperature register */
+#define TI_TMP108_REG_CONF			0x01   /** Configuration register */
```
This change is not needed

mmahadevan108 commented on drivers/sensor/tmp108/tmp108.h:
```diff
@@ -13,33 +14,30 @@
-#define TI_TMP108_CONF_POL	0x0080 /** Alert pin Polarity configuration bit */
-#define TI_TMP108_CONF_TM	0x0400 /** Thermostat mode setting bit */
-#define TI_TMP108_CONF_HYS1	0x0020 /** Temperature hysteresis config 1 bit  */
-#define TI_TMP108_CONF_HYS0	0x0010 /** Temperature hysteresis config 2 bit */
```
Does this need to be deleted?

henrikbrixandersen commented:
I'd go for two separate drivers given the details presented so far.

mbolivar-nordic commented:
I haven't looked at the details but either alternative (1 driver that supports 2 compatibles or 2 drivers) are fine with me depending on whatever makes the most sense to the sensor folks.

jjjtmobile committed:
drivers: sensor: add ams AS621x Temperature Sensor driver support

Added support for the AMS AS621x series of temperature sensors as a
variant of the TI TMP108 temperature sensor.

Signed-off-by: Jared Baumann <jared.baumann8@t-mobile.com>
Signed-off-by: James Johnson <james.johnson672@t-mobile.com>

mbolivar-nordic commented on drivers/sensor/tmp108/Kconfig:
```diff
@@ -1,16 +1,17 @@
+	  the AMS AS621.
 
 if TMP108
-
```
Please don't make unnecessary whitespace changes like this. They clutter up the diff for reviewers and violate the "a commit does one thing" principle. In fact, they are the canonical example of doing two things :).

Slide 14: https://static.sched.com/hosted_files/zephyr2022/a3/HOWTO_Get_Your_Zephyr_Patches_Merged_ZDS_2022.pdf

catch-twenty-two replied:
sorry about that, will do

uLipe commented on drivers/sensor/tmp108/tmp108.c:
```diff
@@ -202,99 +192,106 @@ static int tmp108_attr_set(const struct device *dev,
 	switch ((int) attr) {
 	case SENSOR_ATTR_HYSTERESIS:
+		if (TI_TMP108_HYSTER_0_C(dev) == TI_TMP108_CONF_NA) {
+			LOG_WRN("AS621x Series lacks Hysterisis setttings");
```
If sensor attribute is not supported, ,it should return -ENOTSUP instead a plain warning

catch-twenty-two replied:
good catch, done

uLipe commented on samples/sensor/tmp108/src/main.c:
```diff
@@ -142,8 +142,14 @@ void main(void)
-		return;
+		printf("warning: tmp108 device not found checking for compatible ams device\n");
+
+		tmp108 = DEVICE_DT_GET_ANY(ams_as6212);
```
Since this is a sample I think is worth to be nit here, I suggest to rename tmp108 variable, it is strange to reuse it to gather another device with a different PN, even if they're compatible.

People who refer samples in general are searching a place to start, so keep code simpler to understand is always a good choice there.

something like `sensor_device` maybe? or declare two device descriptors, one to each sensor type.

catch-twenty-two replied:
done

uLipe reviewed (changes requested):
Some minor nits, otherwise, nice work!

catch-twenty-two commented:
@MaureenHelm and @mbolivar-nordic clean up and requested changes have been implemented, please review when you have some time. Thank you

catch-twenty-two commented:
> @MaureenHelm and @mbolivar-nordic clean up and requested changes have been implemented, please review when you have some time. Thank you

@MaureenHelm @gmarull @mbolivar-nordic ping

cfriedt reviewed (dismissed):
LGTM

MaureenHelm reviewed (commented):
Please address the CI failures. We had a lot of CI issues last week in main, so you might just need to rebase.

It would also be nice to split the sample changes to a separate commit.

Otherwise, the code changes LGTM

DhruvaG2000 reviewed (changes requested):
Code LGTM,
Please rebase and check if the CI errors go away.

stephanosio commented:
> Please rebase and check if the CI errors go away.

No, the CI failures here are due to an invalid change in this PR.

MaureenHelm commented:
@uLipe please revisit

uLipe reviewed (approved):
LGTM
